internal/manifests/deployment/deployment.go

```diff
@@ -0,0 +1,182 @@
+package deployment
+
+import (
+	"context"
+
+	"github.com/ViaQ/logerr/kverrors"
+	"github.com/ViaQ/logerr/log"
+	"github.com/openshift/elasticsearch-operator/internal/manifests/status"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// CompareFunc is the type for functions that compare two deployments.
+// Return true if two deployment are equal.
+type CompareFunc func(current, desired *appsv1.Deployment) bool
+
+// MutateFunc is the type for functions that mutate the current deployment
+// by applying the values from the desired deployment.
+type MutateFunc func(current, desired *appsv1.Deployment)
+
+// Get returns the k8s deployment for the given object key or an error.
+func Get(ctx context.Context, c client.Client, key client.ObjectKey) (*appsv1.Deployment, error) {
+	dpl := New(key.Name, key.Namespace, nil, 1).Build()
+
+	if err := c.Get(ctx, key, dpl); err != nil {
+		return dpl, kverrors.Wrap(err, "failed to get deployment",
+			"name", dpl.Name,
+			"namespace", dpl.Namespace,
+		)
+	}
+
+	return dpl, nil
+}
+
+// Create will create the given deployment on the api server or return an error on failure
+func Create(ctx context.Context, c client.Client, dpl *appsv1.Deployment) (status.OperationResultType, error) {
+	err := c.Create(ctx, dpl)
+	if err == nil {
+		return status.OperationResultCreated, nil
+	}
+
+	if apierrors.IsAlreadyExists(kverrors.Root(err)) {
+		return status.OperationResultNone, nil
+	}
+
+	return status.OperationResultNone, kverrors.Wrap(err, "failed to create deployment",
+		"name", dpl.Name,
+		"namespace", dpl.Namespace,
+	)
+}
+
+// Update will update an existing deployment if compare func returns true or else leave it unchanged
+func Update(ctx context.Context, c client.Client, dpl *appsv1.Deployment, cmp CompareFunc, mutate MutateFunc) (status.OperationResultType, error) {
+	current := dpl.DeepCopy()
+	key := client.ObjectKey{Name: current.Name, Namespace: current.Namespace}
+	err := c.Get(ctx, key, current)
+	if err != nil {
+		return status.OperationResultNone, kverrors.Wrap(err, "failed to get deployment",
+			"name", current.Name,
+			"namespace", current.Namespace,
+		)
+	}
+
+	if !cmp(current, dpl) {
```

Comment:
can we use a word instead of just "cmp"?

Reply:
Right, this is actually not a compare it is an equality check. Thus, renamed all `cmp` to `equal` and corresponding `CompareFunc` types to `EqualityFunc`. This brings the terminology close to apimachinery's `semantic.Equality.DeepEqual`.